### DISCUSSION TITLE
I am thinking of building a service for more convenient package installing

### DISCUSSION THREAD

**Discussion by CNSeniorious000** [pyodide/pyodide#5004] [Ideas]
I believe there is a significant issue with the current Python package installation ecosystem in comparison to JavaScript's.

Installing from PyPI needs at least 2 fetches: one for resolving dependencies and another for downloading files (from `.whl`). If it have dependencies, more fetches are needed. And if the first satisfied resolution have incompatible dependencies, even more fetches may be needed.

I think this can be done in the server side.

I envision a package delivery service that manages the resolving and bundling steps, simplifying the user experience.

# Motivation

PyPI is really old-fashioned and conservative.

1. There is no point downloading such a large JSON file just to obtain a list of all versions for a package (e.g., the `/simple/pydantic-core` is over 5MB). This issue is particularly problematic for Pyodide.
2. There are times when we only need a specific subfolder from a large utility package. Requiring developers to createnamespace packages and developing within a monorepo is unreasonable. A package deliver system should handle this, allowing packagedevelopers to configure this behavior in their `pyproject.toml` if desired.
3. `README.md` files don't provide much value for downstream users, yet they are included in the `whl` if I remember correctly. Large readme files from packages such as `fastapi` should be excluded from downloads by default in most cases.
4. Extensive docstrings are usually not useful for users in production environment either.

# Draft

For convenience, let’s temporarily call this service `pydelivr`. Similar to how `jsdelivr` wraps around `npm`, the vision for `pydelivr` is as follows (let’s also name its SDK `pydelivr`):

1. `pydelivr.install("pyodide-py/console", "0.26.2")` will only download the `console` folder of the `pyodide-py` package, in a single fetch.
2. `pydelivr.install([a, b, c])` will receive all the files from those packages in a single fetch.
3. `pydelivr.install(..., without_metadata=True)` will exclude the metadata such as `README` files and licenses, which can sometimes be larger than the package itself.
4. `pydelivr.install(..., minify=True)` will strip all comments and docstrings from sources files.
5. Since the SDK is only a downloader, is can be very lightweight (like <30 lines) and not necessary. The endpoints receives requirements from URL queryString and respond with wheel/tarball/JSON with flatten file structure.

---

Currently there are so much package managers for Python (poetry, pdm, uv, rye, ...), which means their developers are wasting time trying to enhance user experience within the limitations of PyPI. Let's create a better PyPI. The python ecosystem deserves it.

How does everybody think about this?

**Comment by ryanking13** [comment]
Thank you for the suggestion. Obviously, the current design of PyPI doesn't take into account use in a browser environment (obviously). In fact, all the package managers in the world were not built with volatile environments like browsers in mind.

So I think it could be a good idea to create a backend service to solve this pain point.

My only concern is that maintaining a package index will be pretty expensive. We're not a company and we're not managing Pyodide full-time, so we don't have the capacity to run the service ourselves. So, if you're interested in running a service, I think you'll either need to find a company to sponsor it or create your own company using Pyodide.

**Reply by CNSeniorious000** [reply]
That make sense. One slightly different perspective, I believe that can be easier than we thought. For example, we don't need to maintain a registry. This can be just a middleware on PyPI. This solution could start off quite simply, like just running micropip on the server. While it might be slow initially, it would still be faster than running it on the client side, and we can enhance it with caching. If I deploy it as a serverless function, the costs could be low and possibly even free. I think I am going to try it a bit. If I have any positive progress, I would post it here~

**Reply by ryanking13** [reply]
Sure, please share it with us if you make some positive progress.

**Comment by CNSeniorious000** [comment]
I happened to tried a different approach today. I used `uv` (a python package manager written in rust) to lock dependencies into an toml file, then parse it in JavaScript, and at last inject pyodide wheels for non-pure-python packages.

I made [a simple demo](https://github.com/CNSeniorious000/uv-for-pyodide) for this:

[![image](https://github.com/user-attachments/assets/023d17ef-50b0-40af-ac1c-ba6b2c9e7b80)](https://uv-for-pyodide.up.railway.app/)

You can paste a valid `pyproject.toml` into it and the resolution result will appear instantly in the right panel.

> [!tip]
> Try resolving huge packages like `jupyter` / `poetry` within a second!

**Comment by ryanking13** [comment]
Cool, thanks for sharing your project @CNSeniorious000. What's your future plan to integrate it with Pyodide or package delivery service?

**Reply by CNSeniorious000** [reply]
This is still a very rough prototype, to be honest.

I came up with the idea just last night when I realized that the information stored in `uv.lock` is so comprehensive that parsing it could potentially replace the resolution step in micropip.

Of course, it's still got a lot of shortcomings - for example, it doesn't follow the `sys_platform` and `platform_system` specifiers in the pyodide environment.

But I'm confident that this approach is viable and correct: It would be more efficient than creating a new framework from scratch or utilizing existing ecosystems.

Next, I plan to share this 'resolution as a service' vision with the uv team, and if they can add some CLI parameters to specify specifiers in pyodide, or even expose an JavaScript binding package, then we can make progress towards this goal.

**Reply by CNSeniorious000** [reply]
Oh, I should make an SDK to use this service with micropip too.

Or this can be done at build-time (I mean, during bundling frontend assets). The SDK should expose a command to download all the wheels and then merge them into a big wheel.

Then no micropip is needed at runtime! `loadPackage('huge-bundled-wheels.whl')` is all you need. Am I right?

**Reply by hoodmane** [reply]
Have you looked at what `pyodide venv` does? I've tried to fix `sys_platform`, `platform_system`, etc and it will also download pure Python wheels from pypi but it will only download emscripten binary packages from jsdelivr.

**Reply by CNSeniorious000** [reply]
Looks like this is the best way for build-time bundling. I forgot to check these pyodide cli apis these days indeed.

I will have a try tomorrow. Thanks!

**Reply by hoodmane** [reply]
The source code for `pyodide venv` is here:
https://github.com/pyodide/pyodide-build/blob/main/pyodide_build/out_of_tree/venv.py

**Reply by CNSeniorious000** [reply]
I encountered `ValueError: No compatible cross-build environment found`

**Reply by ryanking13** [reply]
> I encountered ValueError: No compatible cross-build environment found

What's the Python and pyodide-build you are using? And what's the output of

```
pyodide xbuildenv search --all
```

?

**Reply by CNSeniorious000** [reply]
```
> pyodide xbuildenv search --all
┌────────────┬────────────┬────────────┬───────────────────────────┬────────────┐
│ Version    │ Python     │ Emscripten │ pyodide-build             │ Compatible │
├────────────┼────────────┼────────────┼───────────────────────────┼────────────┤
│ 0.27.0a2   │ 3.12.1     │ 3.1.58     │ 0.26.0 -                  │ No         │
│ 0.26.2     │ 3.12.1     │ 3.1.58     │ 0.26.0 -                  │ No         │
│ 0.26.1     │ 3.12.1     │ 3.1.58     │ 0.26.0 -                  │ No         │
│ 0.26.0     │ 3.12.1     │ 3.1.58     │ 0.26.0 -                  │ No         │
│ 0.26.0a6   │ 3.12.1     │ 3.1.58     │ 0.26.0a6 - 0.26.0a6       │ No         │
│ 0.26.0a5   │ 3.12.1     │ 3.1.58     │ 0.26.0a5 - 0.26.0a5       │ No         │
│ 0.26.0a4   │ 3.12.1     │ 3.1.52     │ 0.26.0a4 - 0.26.0a4       │ No         │
│ 0.26.0a3   │ 3.12.1     │ 3.1.52     │ 0.26.0a3 - 0.26.0a3       │ No         │
│ 0.26.0a2   │ 3.12.1     │ 3.1.52     │ 0.26.0a2 - 0.26.0a2       │ No         │
│ 0.26.0a1   │ 3.12.1     │ 3.1.52     │ 0.26.0a1 - 0.26.0a1       │ No         │
│ 0.25.1     │ 3.11.3     │ 3.1.46     │ 0.25.1 - 0.25.1           │ No         │
│ 0.25.0     │ 3.11.3     │ 3.1.46     │ 0.25.0 - 0.25.0           │ No         │
└────────────┴────────────┴────────────┴───────────────────────────┴────────────┘
```

Note that I didn't install any things other than `pyodide-cli` and `pyodide-build`. Do I need to install some build environment? I just want to download wheels so I think that shouldn't be necessary to install them.

I installed `pyodide-cli` and `pyodide-build` using a tool like `pipx` to isolated them with other packages:

```
uv tool install pyodide-cli --with pyodide-build --python 3.13
```

so I believe I am using latest versions:

```
Pyodide CLI Version: 0.2.4
```

<details><summary>Details</summary>
<pre>
 + annotated-types==0.7.0
 + anyio==4.6.0
 + auditwheel-emscripten==0.0.16
 + build==1.2.2.post1
 + certifi==2024.8.30
 + charset-normalizer==3.4.0
 + click==8.1.7
 + cmake==3.30.4
 + colorama==0.4.6
 + distlib==0.3.9
 + filelock==3.16.1
 + h11==0.14.0
 + httpcore==1.0.6
 + httpx==0.27.2
 + idna==3.10
 + leb128==1.0.8
 + markdown-it-py==3.0.0
 + mdurl==0.1.2
 + packaging==24.1
 + platformdirs==4.3.6
 + pydantic==2.9.2
 + pydantic-core==2.23.4
 + pygments==2.18.0
 + pyodide-build==0.29.0
 + pyodide-cli==0.2.4
 + pyodide-lock==0.1.0a7
 + pyproject-hooks==1.2.0
 + requests==2.32.3
 + resolvelib==1.0.1
 + rich==13.9.2
 + ruamel-yaml==0.18.6
 + shellingham==1.5.4
 + sniffio==1.3.1
 + typer==0.12.5
 + typing-extensions==4.12.2
 + unearth==0.17.2
 + urllib3==2.2.3
 + virtualenv==20.26.6
 + wheel==0.44.0
</pre>
</details>

**Reply by ryanking13** [reply]
@CNSeniorious000 There is no Pyodide release with Python 3.13. The latest release 0.26.X (and the next release 0.27) will use Python 3.12. So I think you need to use Python 3.12.
